### ISSUE TITLE
Akka Test Kit Equivalent

### ISSUE THREAD

**Issue by nicholasham** [AsynkronIT/protoactor-dotnet#336]
Trying to work out how you would go about testing a simple actor that sends a message back on the parent context. It would be good if you had example of how testing is done with middle-ware or some kind of TestKit like Akka.

Trying to test this simple actor and not sure how to go about it.

` public class DeviceLookupActor : IActor
    {
        private readonly IDeviceLookup _deviceLookup;

        public DeviceLookupActor(IDeviceLookup deviceLookup)
        {
            _deviceLookup = deviceLookup;
        }

        public async Task ReceiveAsync(IContext context)
        {
            switch (context.Message)
            {
                case Started _:
                    break;
                case LookupDevice _:
                    await _deviceLookup.LookupAsync().MatchAsync(
                        device => context.Parent.Tell(new DeviceFound(device)),
                        () => context.Parent.Tell(new DeviceNotFound()));
                    break;
            }
        }

        public class LookupDevice
        {
        }

        public class DeviceFound
        {
            public DeviceFound(Device device)
            {
                Device = device;
            }

            public Device Device { get; }
        }

        public class DeviceNotFound
        {
        }
    }`

**Comment by nicholasham** [comment]
Just to clarify writing a interaction based unit test is not a problem but if I wanted to test it through the actor system runtime then it's not clear if there is a simple way to do this.

**Comment by rogeralsing** [comment]
Our goal is to have enough injection points so that this can be easily done w/o having a special testkit library for it.

We do however have the test mailbox we use in our internal tests, which might be useful to extract.
This lets you write synchronous code and then inspect what has passed through it.

But the general idea is that it should be done with middleware.
I will create an example for it

**Comment by rogeralsing** [comment]
Just a quick note:

```
                        device => context.Parent.Tell(new DeviceFound(device)),
                        () => context.Parent.Tell(new DeviceNotFound()));
```

In order to leverage outgoing middleware, you need to use `context.Tell(target, message)`
This way, the message is passed through the current actors outgoing pipeline.

**Comment by nicholasham** [comment]
Thanks for the clarification regarding the outgoing pipeline and a testing example would be useful.

**Comment by cpx86** [comment]
A test kit was introduced by #469 - should we close this?

**Comment by alexeyzimarev** [comment]
Yeah, I think this issue is only still open because we didn't go through old issues :)

**Comment by alexeyzimarev** [comment]
So, it's here https://github.com/AsynkronIT/protoactor-dotnet/tree/dev/src/Proto.TestKit

Needs docs (as everything else though).
